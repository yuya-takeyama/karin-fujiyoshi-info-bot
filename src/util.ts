import { APIGatewayProxyResult } from 'aws-lambda';

const baseUrl = 'http://www.keyakizaka46.com';
export const path = (path: string) => baseUrl + path;

export const ensureError = (err: any): Error => {
  if (err instanceof Error) {
    return err;
  } else if (err && typeof err.message === 'string') {
    return new Error(err.message);
  } else {
    return new Error(err);
  }
};

export const atomSuccessResponse = (atom: string): APIGatewayProxyResult => {
  return {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/atom+xml',
    },
    body: atom,
  };
};

export const errorToResponse = (error: Error): APIGatewayProxyResult => {
  return {
    statusCode: 500,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      message: error.message,
    }),
  };
};
