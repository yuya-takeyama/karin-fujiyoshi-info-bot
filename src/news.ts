import { JSDOM } from 'jsdom';
import striptags from 'striptags';
import { Feed } from 'feed';
import { parseFromTimeZone } from 'date-fns-timezone';
import { path } from './util';
import fetch from 'node-fetch';

type NewsItem = {
  title: string;
  url: string;
  date: Date;
};

type AtomOptions = {
  updated: Date;
};

export const fetchNewsPage = async (): Promise<string> => {
  const page = await fetch(
    path('/s/k46o/news/list?ima=0000&list%5B%5D=47&reverse=ON'),
  );
  return page.text();
};

export const newsHtmlToAtom = (html: string, options: AtomOptions): string => {
  const newsItems = htmlToNewsItems(html);

  return newsItemsToAtom(newsItems, options);
};

const newsItemsToAtom = (
  newsItems: NewsItem[],
  options: AtomOptions,
): string => {
  const feed = new Feed({
    title: '欅坂46公式サイト 藤吉夏鈴ニュース',
    description: '「坂道シリーズ」第2弾　欅坂46',
    feed: path('/'),
    id: path('/'),
    link: path('/'),
    feedLinks: {},
    copyright: 'Seed & Flower LLC',
    updated: options.updated,
  });
  for (const newsItem of newsItems) {
    feed.addItem({
      title: newsItem.title,
      link: newsItem.url,
      date: newsItem.date,
    });
  }
  return feed.atom1();
};

const htmlToNewsItems = (html: string): NewsItem[] => {
  const dom = new JSDOM(html);
  const elems = dom.window.document.querySelectorAll('.box-news ul li');

  const newsItems: NewsItem[] = [];
  for (const elem of elems) {
    const dateElem = elem.querySelector('.date');
    const linkElem = elem.querySelector('.text a');

    if (linkElem && dateElem) {
      const date = parseFromTimeZone(striptags(dateElem.innerHTML), {
        timeZone: 'Asia/Tokyo',
      });
      const link = linkElem.getAttribute('href');
      const title = striptags(linkElem.innerHTML);

      if (link && date) {
        const url = path(link);
        newsItems.push({
          title,
          url,
          date,
        });
      }
    }
  }

  return newsItems;
};
