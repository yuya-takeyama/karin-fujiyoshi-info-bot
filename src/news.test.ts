import { join } from 'path';
import { readFile as origReadFile } from 'fs';
import { promisify } from 'util';
import { newsHtmlToAtom } from './news';

const readFile = promisify(origReadFile);

const newsFixturesDir = join(__dirname, 'fixtures', 'news');

describe('#newsHtmlToAtom', () => {
  it('returns an atom Feed', async () => {
    const html = (
      await readFile(join(newsFixturesDir, 'news.html'))
    ).toString();
    expect(
      newsHtmlToAtom(html, { updated: new Date(2020, 1, 1, 9, 0, 0) }),
    ).toMatchSnapshot();
  });
});
