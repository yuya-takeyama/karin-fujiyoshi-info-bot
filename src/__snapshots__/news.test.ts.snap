// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#newsHtmlToAtom returns an atom Feed 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<feed xmlns=\\"http://www.w3.org/2005/Atom\\">
    <id>http://www.keyakizaka46.com/</id>
    <title>欅坂46公式サイト 藤吉夏鈴ニュース</title>
    <updated>2020-02-01T00:00:00.000Z</updated>
    <generator>https://github.com/jpmonette/feed</generator>
    <link rel=\\"alternate\\" href=\\"http://www.keyakizaka46.com/\\"/>
    <link rel=\\"self\\" href=\\"http://www.keyakizaka46.com/\\"/>
    <subtitle>「坂道シリーズ」第2弾　欅坂46</subtitle>
    <rights>Seed &amp; Flower LLC</rights>
    <entry>
        <title type=\\"html\\"><![CDATA[「BUBKA」11月号(9月30日(水)発売)の表紙に渡邉理佐、中面に小林由依、菅井友香、土生瑞穂、藤吉夏鈴、山﨑天、大園玲が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01772?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01772?ima=0000\\">
        </link>
        <updated>2020-09-29T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[本日6月29日(月)、欅坂46 新二期生が「欅坂46メッセージ」と個別ブログを開始!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/O00203?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/O00203?ima=0000\\">
        </link>
        <updated>2020-06-28T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[6月29日(月)より、欅坂46 新二期生が個別ブログと「欅坂46メッセージ」を開始! ]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/O00204?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/O00204?ima=0000\\">
        </link>
        <updated>2020-06-18T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「EX大衆」5月号(4月15日(水)発売)の表紙に小林由依、森田ひかる、中面に上村莉菜、原田葵、武元唯衣、藤吉夏鈴、山﨑天が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01635?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01635?ima=0000\\">
        </link>
        <updated>2020-04-14T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[2020年2月29日(土)「第30回東京ガールズコレクション2020　SPRING/SUMMER」 に土生瑞穂、渡辺梨加、渡邉理佐、藤吉夏鈴、森田ひかる、山﨑天のモデル出演が決定!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/E00459?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/E00459?ima=0000\\">
        </link>
        <updated>2020-01-30T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「Top Yell NEO 2019-2020」(1月8日(水)発売)の表紙・巻頭に井上梨名、田村保乃、藤吉夏鈴、中面に尾関梨香、武元唯衣、山﨑天が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01540?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01540?ima=0000\\">
        </link>
        <updated>2020-01-07T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「BUBKA」2月号(12月28日(土)発売)に藤吉夏鈴、山﨑天が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01535?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01535?ima=0000\\">
        </link>
        <updated>2019-12-27T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「BOMB」1月号(12月9日(月)発売)に井上梨名、藤吉夏鈴が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01508?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01508?ima=0000\\">
        </link>
        <updated>2019-12-08T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「BIG ONE GIRLS」12月号(11月30日(土)発売)表紙・巻頭に武元唯衣、藤吉夏鈴が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01492?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01492?ima=0000\\">
        </link>
        <updated>2019-11-29T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「B.L.T.」2020年1月号(11月22日(金)発売)の表紙に二期生全員、中面に上村莉菜、原田葵、小池美波が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01483?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01483?ima=0000\\">
        </link>
        <updated>2019-11-21T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「EX大衆」11月号(10月15日(火)発売)の表紙に小林由依、中面に藤吉夏鈴、インタビューで土生瑞穂、井上梨名×武元唯衣が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01459?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01459?ima=0000\\">
        </link>
        <updated>2019-10-14T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「blt graph.」vol.47(9月11日(水)発売)に小林由依、藤吉夏鈴が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01446?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01446?ima=0000\\">
        </link>
        <updated>2019-09-10T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「EX大衆」9月号(8月16日(金)発売)に小池美波、原田葵、藤吉夏鈴×松田里奈が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01428?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01428?ima=0000\\">
        </link>
        <updated>2019-08-15T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「BUBKA」9月号(7月31日(水)発売)の尾関梨香、山﨑天が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01417?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01417?ima=0000\\">
        </link>
        <updated>2019-07-30T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「週刊少年サンデー」34号(7月24日(水)発売)表紙・巻頭に藤吉夏鈴、森田ひかるが登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01410?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01410?ima=0000\\">
        </link>
        <updated>2019-07-23T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「Platinum FLASH」Vol.10(7月19日(金)発売)の裏表紙に井上梨名、田村保乃、藤吉夏鈴が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01405?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01405?ima=0000\\">
        </link>
        <updated>2019-07-18T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[7月31日(水)～8月11日(日)新感覚ホラーアトラクション「ザンビ THE ROOM 最後の選択」出演の追加メンバー決定!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/E00410?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/E00410?ima=0000\\">
        </link>
        <updated>2019-07-17T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「BUBKA」8月号(6月28日(金)発売)の表紙に欅坂46二期生が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01391?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01391?ima=0000\\">
        </link>
        <updated>2019-06-27T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「MARQUEE」Vol.133(6月15日(土)発売)の表紙に田村保乃と藤吉夏鈴登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01364?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01364?ima=0000\\">
        </link>
        <updated>2019-06-09T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「週刊少年マガジン」27号(6月5日(水)発売)の巻頭に藤吉夏鈴が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01363?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01363?ima=0000\\">
        </link>
        <updated>2019-06-04T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[6月3日(月)～6月7日(金)欅坂46 二期生「個別握手会お礼SHOWROOM」配信!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01366?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01366?ima=0000\\">
        </link>
        <updated>2019-06-02T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[5月23日(木)25:59～読売テレビ「音力-ONCHIKA-」に欅坂46二期生が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01354?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01354?ima=0000\\">
        </link>
        <updated>2019-05-21T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「欅坂46メッセージ」に欅坂46 二期生メンバーが追加! ]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/O00158?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/O00158?ima=0000\\">
        </link>
        <updated>2019-05-13T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[欅坂46二期生　8thシングル「黒い羊」発売記念　個別握手会参加決定!第7次～第9次応募受付発表!!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/S00370?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/S00370?ima=0000\\">
        </link>
        <updated>2019-05-12T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「BOMB」6月号(5月9日(木)発売)に武元唯衣、藤吉夏鈴、森田ひかるが登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01347?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01347?ima=0000\\">
        </link>
        <updated>2019-05-08T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「EX大衆」5月号(4月15日(月)発売)に藤吉夏鈴、井上梨名×関有美子×松平璃子×山﨑天が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01332?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01332?ima=0000\\">
        </link>
        <updated>2019-04-14T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「日経エンタテインメント!」2019年5月号(4月4日(木)発売)表紙に石森虹花、織田奈那、小池美波、小林由依、佐藤詩織、菅井友香、土生瑞穂、守屋茜、渡辺梨加、渡邉理佐が登場!封入のインタビューBOOKには井上梨名、関有美子、武元唯衣、田村保乃、藤吉夏鈴、松田里奈、松平璃子、森田ひかる、山﨑天も登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01324?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01324?ima=0000\\">
        </link>
        <updated>2019-04-03T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「B.L.T.」2019年5月号(3月23日(土)発売)表紙に日向坂46加藤史帆、齊藤京子、小坂菜緒、中面にメンバー全員が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01302?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01302?ima=0000\\">
        </link>
        <updated>2019-03-22T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「Platinum FLASH」Vol.9(3月8日(金)発売)に欅坂46 二期生が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01283?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01283?ima=0000\\">
        </link>
        <updated>2019-03-07T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[2月26日(火)公開の「TOKYO POP LINE」に藤吉夏鈴、山﨑天が登場!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01259?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01259?ima=0000\\">
        </link>
        <updated>2019-02-25T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「B.L.T.」2019年4月号(2月23日(土)発売)に土生瑞穂が登場!欅坂46 二期生PERFECT BOOK 欅図鑑も掲載!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01248?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01248?ima=0000\\">
        </link>
        <updated>2019-02-22T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[「BOMB」3月号(2月9日(土)発売)表紙・巻頭に上村莉菜×尾関梨香×菅井友香、中面に小池美波×土生瑞穂、齋藤冬優花×佐藤詩織、守屋茜、上村ひなのが登場!欅坂46 二期生名鑑も掲載!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/M01229?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/M01229?ima=0000\\">
        </link>
        <updated>2019-02-08T15:00:00.000Z</updated>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[2月27日(水)リリース、8thシングル収録内容発表!]]></title>
        <id>http://www.keyakizaka46.com/s/k46o/news/detail/R00088?ima=0000</id>
        <link href=\\"http://www.keyakizaka46.com/s/k46o/news/detail/R00088?ima=0000\\">
        </link>
        <updated>2019-01-30T15:00:00.000Z</updated>
    </entry>
</feed>"
`;
